// @ziplayer/ytexecplug/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.YTexec = void 0;
exports.webStreamToNodeStream = webStreamToNodeStream;
const ziplayer_1 = require("ziplayer");
const stream_1 = require("stream");
const youtube_dl_exec_1 = __importDefault(require("youtube-dl-exec"));
/**
 * Converts a Web ReadableStream to a Node.js Readable stream
 */
function webStreamToNodeStream(webStream) {
    const nodeStream = new stream_1.Readable({
        read() {
            // This will be handled by the Web Stream reader
        },
    });
    // Create a reader from the Web Stream
    const reader = webStream.getReader();
    // Read chunks and push to Node.js stream
    const pump = async () => {
        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done) {
                    nodeStream.push(null); // End the stream
                    break;
                }
                nodeStream.push(Buffer.from(value));
            }
        }
        catch (error) {
            nodeStream.destroy(error);
        }
    };
    // Start pumping data
    pump();
    return nodeStream;
}
function extractVideoId(input) {
    try {
        const u = new URL(input);
        const allowedShortHosts = ["youtu.be"];
        const allowedLongHosts = ["youtube.com", "www.youtube.com", "music.youtube.com", "m.youtube.com"];
        if (allowedShortHosts.includes(u.hostname)) {
            return u.pathname.split("/").filter(Boolean)[0] || null;
        }
        if (allowedLongHosts.includes(u.hostname)) {
            // watch?v=, shorts/, embed/
            if (u.searchParams.get("v"))
                return u.searchParams.get("v");
            const path = u.pathname;
            if (path.startsWith("/shorts/"))
                return path.replace("/shorts/", "");
            if (path.startsWith("/embed/"))
                return path.replace("/embed/", "");
        }
        return null;
    }
    catch {
        return null;
    }
}
async function getYoutubeStream(url) {
    const URL = "https://youtu.be/" + extractVideoId(url);
    const info = await (0, youtube_dl_exec_1.default)(URL, {
        dumpSingleJson: true,
        noCheckCertificates: true,
        noWarnings: true,
        preferFreeFormats: true,
        format: "bestaudio/best",
        addHeader: ["referer:youtube.com", "user-agent:googlebot"],
    });
    const videourl = typeof info === "object" ? info?.url : info;
    if (!videourl) {
        return null;
    }
    return videourl;
}
class YTexec extends ziplayer_1.BasePlugin {
    constructor() {
        super(...arguments);
        this.name = "YTexec";
        this.version = "1.0.0";
    }
    canHandle(query) {
        const q = (query || "").trim().toLowerCase();
        const isUrl = q.startsWith("http://") || q.startsWith("https://");
        if (isUrl) {
            try {
                const parsed = new URL(query);
                const allowedHosts = ["youtube.com", "www.youtube.com", "music.youtube.com", "youtu.be", "www.youtu.be"];
                return allowedHosts.includes(parsed.hostname.toLowerCase());
            }
            catch (e) {
                return false;
            }
        }
        return false;
    }
    async search(query, requestedBy) {
        return { tracks: [] };
    }
    async getStream(track) {
        try {
            const youtubeUrl = await getYoutubeStream(track.url);
            if (!youtubeUrl) {
                throw new Error("Failed to get YouTube stream URL");
            }
            const response = await fetch(youtubeUrl);
            if (!response.ok || !response.body) {
                throw new Error("Failed to fetch YouTube stream");
            }
            const stream = webStreamToNodeStream(response.body);
            return {
                stream,
                type: "arbitrary",
                metadata: track.metadata,
            };
        }
        catch (error) {
            throw new Error(`Failed to get YouTube stream: ${error}`);
        }
    }
}
exports.YTexec = YTexec;
