// @ziplayer/ytexecplug/dist/YouTubePlugin.js
"use strict";
// import { BasePlugin, Track, SearchResult, StreamInfo, Player } from "ziplayer";
// export interface PluginOptions {
// 	player: Player;
// 	debug?: boolean;
// }
// /**
//  * A plugin for handling YouTube audio content including videos, playlists, and search functionality.
//  *
//  * This plugin provides comprehensive support for:
//  * - YouTube video URLs (youtube.com, youtu.be, music.youtube.com)
//  * - YouTube playlist URLs and dynamic mixes
//  * - YouTube search queries
//  * - Audio stream extraction from YouTube videos
//  * - Related track recommendations
//  *
//  * @example
//  * const youtubePlugin = new YouTubePlugin();
//  *
//  * // Add to PlayerManager
//  * const manager = new PlayerManager({
//  *   plugins: [youtubePlugin]
//  * });
//  *
//  * // Search for videos
//  * const result = await youtubePlugin.search("Never Gonna Give You Up", "user123");
//  *
//  * // Get audio stream
//  * const stream = await youtubePlugin.getStream(result.tracks[0]);
//  *
//  * @since 1.0.0
//  */
// export class YouTubePlugin extends BasePlugin {
// 	name = "youtube";
// 	version = "1.0.0";
// 	private ready: Promise<void>;
// 	private player: Player | undefined;
// 	private options: PluginOptions;
// 	/**
// 	 * Creates a new YouTubePlugin instance.
// 	 *
// 	 * The plugin will automatically initialize YouTube clients for both video playback
// 	 * and search functionality. Initialization is asynchronous and handled internally.
// 	 *
// 	 * @example
// 	 * const plugin = new YouTubePlugin();
// 	 * // Plugin is ready to use after initialization completes
// 	 */
// 	constructor(options: PluginOptions) {
// 		super();
// 		this.player = options?.player ?? undefined;
// 		this.options = options ?? {};
// 		this.ready = this.init();
// 	}
// 	private async init(): Promise<void> {}
// 	private debug(message?: any, ...optionalParams: any[]): void {
// 		if (this.options?.debug && this?.player && this.player?.listenerCount("debug") > 0) {
// 			this.player.emit("debug", `[YouTubePlugin] ${message}`, ...optionalParams);
// 		}
// 	}
// 	// Build a Track from various YouTube object shapes (search item, playlist item, watch_next feed, basic_info, info)
// 	private buildTrack(raw: any, requestedBy: string, extra?: { playlist?: string }): Track {
// 		const pickFirst = (...vals: any[]) => vals.find((v) => v !== undefined && v !== null && v !== "");
// 		// Try to resolve from multiple common shapes
// 		const id = pickFirst(
// 			raw?.id,
// 			raw?.video_id,
// 			raw?.videoId,
// 			raw?.content_id,
// 			raw?.identifier,
// 			raw?.basic_info?.id,
// 			raw?.basic_info?.video_id,
// 			raw?.basic_info?.videoId,
// 			raw?.basic_info?.content_id,
// 		);
// 		const title = pickFirst(
// 			raw?.metadata?.title?.text,
// 			raw?.title?.text,
// 			raw?.title,
// 			raw?.headline,
// 			raw?.basic_info?.title,
// 			"Unknown title",
// 		);
// 		const durationValue = pickFirst(
// 			raw?.length_seconds,
// 			raw?.duration?.seconds,
// 			raw?.duration?.text,
// 			raw?.duration,
// 			raw?.length_text,
// 			raw?.basic_info?.duration,
// 		);
// 		const duration = Number(toSeconds(durationValue)) || 0;
// 		const thumb = pickFirst(
// 			raw?.thumbnails?.[0]?.url,
// 			raw?.thumbnail?.[0]?.url,
// 			raw?.thumbnail?.url,
// 			raw?.thumbnail?.thumbnails?.[0]?.url,
// 			raw?.content_image?.image?.[0]?.url,
// 			raw?.basic_info?.thumbnail?.[0]?.url,
// 			raw?.basic_info?.thumbnail?.[raw?.basic_info?.thumbnail?.length - 1]?.url,
// 			raw?.thumbnails?.[raw?.thumbnails?.length - 1]?.url,
// 		);
// 		const author = pickFirst(raw?.author?.name, raw?.author, raw?.channel?.name, raw?.owner?.name, raw?.basic_info?.author);
// 		const views = pickFirst(
// 			raw?.view_count,
// 			raw?.views,
// 			raw?.short_view_count,
// 			raw?.stats?.view_count,
// 			raw?.basic_info?.view_count,
// 		);
// 		const url = pickFirst(raw?.url, id ? `https://www.youtube.com/watch?v=${id}` : undefined);
// 		this.debug("Track build:", {
// 			id: String(id),
// 			title: String(title),
// 			url: String(url),
// 			duration,
// 			thumbnail: thumb,
// 			requestedBy,
// 			source: this.name,
// 		});
// 		return {
// 			id: String(id),
// 			title: String(title),
// 			url: String(url),
// 			duration,
// 			thumbnail: thumb,
// 			requestedBy,
// 			source: this.name,
// 			metadata: {
// 				author,
// 				views,
// 				...(extra?.playlist ? { playlist: extra.playlist } : {}),
// 			},
// 		} as Track;
// 	}
// 	/**
// 	 * Determines if this plugin can handle the given query.
// 	 *
// 	 * @param query - The search query or URL to check
// 	 * @returns `true` if the plugin can handle the query, `false` otherwise
// 	 *
// 	 * @example
// 	 * plugin.canHandle("https://www.youtube.com/watch?v=dQw4w9WgXcQ"); // true
// 	 * plugin.canHandle("Never Gonna Give You Up"); // true
// 	 * plugin.canHandle("spotify:track:123"); // false
// 	 */
// 	canHandle(query: string): boolean {
// 		const q = (query || "").trim().toLowerCase();
// 		const isUrl = q.startsWith("http://") || q.startsWith("https://");
// 		if (isUrl) {
// 			try {
// 				const parsed = new URL(query);
// 				const allowedHosts = ["youtube.com", "www.youtube.com", "music.youtube.com", "youtu.be", "www.youtu.be"];
// 				return allowedHosts.includes(parsed.hostname.toLowerCase());
// 			} catch (e) {
// 				return false;
// 			}
// 		}
// 		// Avoid intercepting explicit patterns for other extractors
// 		if (q.startsWith("tts:") || q.startsWith("say ")) return false;
// 		if (q.startsWith("spotify:") || q.includes("open.spotify.com")) return false;
// 		if (q.includes("soundcloud")) return false;
// 		// Treat remaining non-URL free text as YouTube-searchable
// 		return true;
// 	}
// 	/**
// 	 * Validates if a URL is a valid YouTube URL.
// 	 *
// 	 * @param url - The URL to validate
// 	 * @returns `true` if the URL is a valid YouTube URL, `false` otherwise
// 	 *
// 	 * @example
// 	 * plugin.validate("https://www.youtube.com/watch?v=dQw4w9WgXcQ"); // true
// 	 * plugin.validate("https://youtu.be/dQw4w9WgXcQ"); // true
// 	 * plugin.validate("https://spotify.com/track/123"); // false
// 	 */
// 	validate(url: string): boolean {
// 		try {
// 			const parsed = new URL(url);
// 			const allowedHosts = ["youtube.com", "www.youtube.com", "music.youtube.com", "youtu.be", "www.youtu.be", "m.youtube.com"];
// 			return allowedHosts.includes(parsed.hostname.toLowerCase());
// 		} catch (e) {
// 			return false;
// 		}
// 	}
// 	/**
// 	 * Retrieves the audio stream for a YouTube track using sabr download.
// 	 *
// 	 * This method extracts the audio stream from a YouTube video using the sabr download
// 	 * method which provides better quality and more reliable streaming.
// 	 *
// 	 * @param track - The Track object to get the stream for
// 	 * @returns A StreamInfo object containing the audio stream and metadata
// 	 * @throws {Error} If the track ID is invalid or stream extraction fails
// 	 *
// 	 * @example
// 	 * const track = { id: "dQw4w9WgXcQ", title: "Never Gonna Give You Up", ... };
// 	 * const streamInfo = await plugin.getStream(track);
// 	 * console.log(streamInfo.type); // "arbitrary"
// 	 * console.log(streamInfo.stream); // Readable stream
// 	 */
// 	async getStream(track: Track): Promise<StreamInfo> {
// 		await this.ready;
// 		const id = this.extractVideoId(track.url) || track.id;
// 		if (!id) throw new Error("Invalid track id");
// 		try {
// 			this.debug("🚀 Attempting sabr download for video ID:", id);
// 			// Use sabr download for better quality and reliability
// 			const { streamResults } = await createSabrStream(id, DEFAULT_SABR_OPTIONS);
// 			const { audioStream, selectedFormats, videoTitle } = streamResults;
// 			this.debug("✅ Sabr download successful, converting Web Stream to Node.js Stream");
// 			// Convert Web Stream to Node.js Readable Stream
// 			const nodeStream = webStreamToNodeStream(audioStream);
// 			this.debug("✅ Stream conversion complete, returning Node.js stream");
// 			// Return the converted Node.js stream
// 			return {
// 				stream: nodeStream,
// 				type: "arbitrary",
// 				metadata: {
// 					...track.metadata,
// 					itag: selectedFormats.audioFormat.itag,
// 					mime: selectedFormats.audioFormat.mimeType,
// 				},
// 			};
// 		} catch (e: any) {
// 			this.debug("⚠️ Sabr download failed, falling back to youtubei.js:", e.message);
// 			// Fallback to original youtubei.js method if sabr download fails
// 			try {
// 				const stream: any = await (this.client as any).download(id, {
// 					type: "audio",
// 					quality: "best",
// 				});
// 				// Check if it's a Web Stream and convert it
// 				this.debug("🔍 Checking stream type:", typeof stream, stream?.constructor?.name);
// 				if (stream && typeof stream.getReader === "function") {
// 					this.debug("🔄 Converting Web Stream to Node.js Stream");
// 					const nodeStream = webStreamToNodeStream(stream);
// 					this.debug("✅ Stream converted successfully");
// 					return {
// 						stream: nodeStream,
// 						type: "arbitrary",
// 						metadata: track.metadata,
// 					};
// 				} else {
// 					this.debug("⚠️ Stream is not a Web Stream or is null");
// 				}
// 				return {
// 					stream,
// 					type: "arbitrary",
// 					metadata: track.metadata,
// 				};
// 			} catch (fallbackError: any) {
// 				try {
// 					const info: any = await (this.client as any).getBasicInfo(id);
// 					// Prefer m4a audio-only formats first
// 					let format: any = info?.chooseFormat?.({
// 						type: "audio",
// 						quality: "best",
// 					});
// 					if (!format && info?.formats?.length) {
// 						const audioOnly = info.formats.filter((f: any) => f.mime_type?.includes("audio"));
// 						audioOnly.sort((a: any, b: any) => (b.bitrate || 0) - (a.bitrate || 0));
// 						format = audioOnly[0];
// 					}
// 					if (!format) throw new Error("No audio format available");
// 					let url: string | undefined = undefined;
// 					if (typeof format.decipher === "function") {
// 						url = format.decipher((this.client as any).session.player);
// 					}
// 					if (!url) url = format.url;
// 					if (!url) throw new Error("No valid URL to decipher");
// 					const res = await fetch(url);
// 					if (!res.ok || !res.body) {
// 						throw new Error(`HTTP ${res.status}`);
// 					}
// 					// Convert Web Stream to Node.js Stream
// 					this.debug("🔄 Converting fetch response Web Stream to Node.js Stream");
// 					const nodeStream = webStreamToNodeStream(res.body);
// 					return {
// 						stream: nodeStream,
// 						type: "arbitrary",
// 						metadata: {
// 							...track.metadata,
// 							itag: format.itag,
// 							mime: format.mime_type,
// 						},
// 					};
// 				} catch (inner: any) {
// 					throw new Error(`Failed to get YouTube stream: ${inner?.message || inner}`);
// 				}
// 			}
// 		}
// 	}
// 	async getFallback(track: Track): Promise<StreamInfo> {
// 		try {
// 			const result = await this.search(track.title, track.requestedBy);
// 			const first = result.tracks[0];
// 			this.debug("Fallback track:", first);
// 			if (!first) throw new Error("No fallback track found");
// 			return await this.getStream(first);
// 		} catch (e: any) {
// 			throw new Error(`YouTube fallback search failed: ${e?.message || e}`);
// 		}
// 	}
// 	private extractVideoId(input: string): string | null {
// 		try {
// 			const u = new URL(input);
// 			const allowedShortHosts = ["youtu.be"];
// 			const allowedLongHosts = ["youtube.com", "www.youtube.com", "music.youtube.com", "m.youtube.com"];
// 			if (allowedShortHosts.includes(u.hostname)) {
// 				return u.pathname.split("/").filter(Boolean)[0] || null;
// 			}
// 			if (allowedLongHosts.includes(u.hostname)) {
// 				// watch?v=, shorts/, embed/
// 				if (u.searchParams.get("v")) return u.searchParams.get("v");
// 				const path = u.pathname;
// 				if (path.startsWith("/shorts/")) return path.replace("/shorts/", "");
// 				if (path.startsWith("/embed/")) return path.replace("/embed/", "");
// 			}
// 			return null;
// 		} catch {
// 			return null;
// 		}
// 	}
// }
// function toSeconds(d: any): number | undefined {
// 	if (typeof d === "number") return d;
// 	if (typeof d === "string") {
// 		// mm:ss or hh:mm:ss
// 		const parts = d.split(":").map(Number);
// 		if (parts.some((n) => Number.isNaN(n))) return undefined;
// 		if (parts.length === 3) return parts[0] * 3600 + parts[1] * 60 + parts[2];
// 		if (parts.length === 2) return parts[0] * 60 + parts[1];
// 		const asNum = Number(d);
// 		return Number.isFinite(asNum) ? asNum : undefined;
// 	}
// 	if (d && typeof d === "object") {
// 		if (typeof (d as any).seconds === "number") return (d as any).seconds;
// 		if (typeof (d as any).milliseconds === "number") return Math.floor((d as any).milliseconds / 1000);
// 	}
// 	return undefined;
// }
