// ziplayer/dist/extensions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionManager = exports.BaseExtension = void 0;
var BaseExtension_1 = require("./BaseExtension");
Object.defineProperty(exports, "BaseExtension", { enumerable: true, get: function () { return BaseExtension_1.BaseExtension; } });
// Extension factory
class ExtensionManager {
    constructor(player, manager) {
        this.player = player;
        this.manager = manager;
        this.extensions = new Map();
        this.extensionContext = Object.freeze({ player, manager });
    }
    debug(message, ...optionalParams) {
        if (this.manager.debugEnabled) {
            this.manager.emit("debug", `[ExtensionManager] ${message}`, ...optionalParams);
        }
    }
    register(extension) {
        if (this.extensions.has(extension.name)) {
            return;
        }
        if (!extension.player) {
            extension.player = this.player;
        }
        this.extensions.set(extension.name, extension);
    }
    unregister(extension) {
        const name = extension.name;
        const result = this.extensions.delete(name);
        if (result) {
            this.invokeExtensionLifecycle(extension, "onDestroy");
        }
        return result;
    }
    destroy() {
        this.debug(`[ExtensionManager] destroying all extensions`);
        for (const extension of this.extensions.values()) {
            this.unregister(extension);
        }
        this.extensions.clear();
    }
    get(name) {
        return this.extensions.get(name);
    }
    getAll() {
        return Array.from(this.extensions.values());
    }
    findExtension(alas) {
        return this.getAll().find((extension) => extension.active(alas));
    }
    clear() {
        this.extensions.clear();
    }
    invokeExtensionLifecycle(extension, hook) {
        if (!extension)
            return;
        const fn = extension[hook];
        if (typeof fn !== "function")
            return;
        try {
            const result = fn.call(extension, this.extensionContext);
            if (result && typeof result.then === "function") {
                result.catch((err) => this.debug(`[Player] Extension ${extension.name} ${hook} error:`, err));
            }
        }
        catch (err) {
            this.debug(`[Player] Extension ${extension.name} ${hook} error:`, err);
        }
    }
    async provideSearch(query, requestedBy) {
        const request = { query, requestedBy };
        for (const extension of this.getAll()) {
            const hook = extension.provideSearch;
            if (typeof hook !== "function")
                continue;
            try {
                const result = await Promise.resolve(hook.call(extension, this.extensionContext, request));
                if (result && Array.isArray(result.tracks) && result.tracks.length > 0) {
                    this.debug(`[Player] Extension ${extension.name} handled search for query: ${query}`);
                    return result;
                }
            }
            catch (err) {
                this.debug(`[Player] Extension ${extension.name} provideSearch error:`, err);
            }
        }
        return null;
    }
    async provideStream(track) {
        const request = { track };
        for (const extension of this.getAll()) {
            const hook = extension.provideStream;
            if (typeof hook !== "function")
                continue;
            try {
                const result = await Promise.resolve(hook.call(extension, this.extensionContext, request));
                if (result && result.stream) {
                    this.debug(`[Player] Extension ${extension.name} provided stream for track: ${track.title}`);
                    return result;
                }
            }
            catch (err) {
                this.debug(`[Player] Extension ${extension.name} provideStream error:`, err);
            }
        }
        return null;
    }
    async BeforePlayHooks(initial) {
        const request = { ...initial };
        const response = {};
        for (const extension of this.getAll()) {
            const hook = extension.beforePlay;
            if (typeof hook !== "function")
                continue;
            try {
                const result = await Promise.resolve(hook.call(extension, this.extensionContext, request));
                if (!result)
                    continue;
                if (result.query !== undefined) {
                    request.query = result.query;
                    response.query = result.query;
                }
                if (result.requestedBy !== undefined) {
                    request.requestedBy = result.requestedBy;
                    response.requestedBy = result.requestedBy;
                }
                if (Array.isArray(result.tracks)) {
                    response.tracks = result.tracks;
                }
                if (typeof result.isPlaylist === "boolean") {
                    response.isPlaylist = result.isPlaylist;
                }
                if (typeof result.success === "boolean") {
                    response.success = result.success;
                }
                if (result.error instanceof Error) {
                    response.error = result.error;
                }
                if (typeof result.handled === "boolean") {
                    response.handled = result.handled;
                    if (result.handled)
                        break;
                }
            }
            catch (err) {
                this.debug(`[Player] Extension ${extension.name} beforePlay error:`, err);
            }
        }
        return { request, response };
    }
    async AfterPlayHooks(payload) {
        if (this.getAll().length === 0)
            return;
        const safeTracks = payload.tracks ? [...payload.tracks] : undefined;
        if (safeTracks) {
            Object.freeze(safeTracks);
        }
        const immutablePayload = Object.freeze({ ...payload, tracks: safeTracks });
        for (const extension of this.getAll()) {
            const hook = extension.afterPlay;
            if (typeof hook !== "function")
                continue;
            try {
                await Promise.resolve(hook.call(extension, this.extensionContext, immutablePayload));
            }
            catch (err) {
                this.debug(`[Player] Extension ${extension.name} afterPlay error:`, err);
            }
        }
    }
}
exports.ExtensionManager = ExtensionManager;
